fvm/evm/offchain/sync/replayer.go

```diff
@@ -0,0 +1,65 @@
+package sync
+
+import (
+	gethTracers "github.com/onflow/go-ethereum/eth/tracers"
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/fvm/evm/events"
+	"github.com/onflow/flow-go/fvm/evm/types"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// Replayer consumes EVM transaction and block
+// events, re-execute EVM transaction and follows EVM chain.
+// this allows using different tracers and storage solutions.
+type Replayer struct {
+	chainID         flow.ChainID
+	rootAddr        flow.Address
+	logger          zerolog.Logger
+	storageProvider types.StorageProvider
+	tracer          *gethTracers.Tracer
+	validateResults bool
+}
+
+// NewReplayer constructs a new Replayer
+func NewReplayer(
+	chainID flow.ChainID,
+	rootAddr flow.Address,
+	sp types.StorageProvider,
+	logger zerolog.Logger,
+	tracer *gethTracers.Tracer,
+	validateResults bool,
+) *Replayer {
+	return &Replayer{
+		chainID:         chainID,
+		rootAddr:        rootAddr,
+		storageProvider: sp,
+		logger:          logger,
+		tracer:          tracer,
+		validateResults: validateResults,
+	}
+}
+
+// OnBlockReceived is called when a new block is received
+// (including all the related transaction executed events)
+func (cr *Replayer) OnBlockReceived(
+	transactionEvents []events.TransactionEventPayload,
+	blockEvent *events.BlockEventPayload,
```

Comment:
Maybe we should drop the pointer on `blockEvent`? It will save us from some `nil` checks throughout the relevant methods.